codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameEncoder.java

```diff
@@ -163,38 +177,77 @@ private static int compressionLevel(int blockSize) {
     }
 
     @Override
+    protected ByteBuf allocateBuffer(ChannelHandlerContext ctx, ByteBuf msg, boolean preferDirect) {
+        return  allocateBuffer(ctx, msg, preferDirect, true);
+    }
+
+    ByteBuf allocateBuffer(ChannelHandlerContext ctx, ByteBuf msg, boolean preferDirect,
+                                     boolean allowEmptyReturn) {
+        int targetBufSize = 0;
+        int remaining = msg.readableBytes() + buffer.readableBytes();
+        while (remaining > 0) {
+            int curSize = Math.min(blockSize, remaining);
+            remaining -= curSize;
+            // calculate the total compressed size of the current block (including header) and add to the total
+            targetBufSize += compressor.maxCompressedLength(curSize) + HEADER_LENGTH;
+        }
+
+        if (targetBufSize > maxEncodeSize || 0 > targetBufSize) {
+            String errorMsg = "requested encode buffer size (%d bytes) exceeds the maximum allowable size (%d bytes)";
+            throw new EncoderException(String.format(errorMsg, targetBufSize, maxEncodeSize));
+        }
+
+        if (allowEmptyReturn && targetBufSize < blockSize) {
+            //would prefer to return null or Unpooled.EMPTY_BUFFER, but MessageToByteEncoder calls buffer.release()
```

Comment:
Huh, I naively assumed that calling `Unpooled.EMPTY_BUFFER.release()` would have thrown an exception after the first release, but I guess that's not the case as you probably wouldn't have made it a constant :) . I ran a quick unit test and indeed it does not error out, so switching to use `Unpooled.EMPTY_BUFFER`.